language/internal/cc/parser/parser.go

```diff
@@ -267,145 +257,153 @@ func parseDefinedExpr(p *parser, op string) (Expr, error) {
+// Drop n tokens from the front of the input stream.
+func (p *parser) dropTokens(n int) {
+	p.tokensLeft = p.tokensLeft[n:]
 }
 
-// parseDirectivesUntil reads tokens and parses directives until shouldStop returns true.
-// It handles main(), #include, and preprocessor blocks, and builds the nested directive structure.
-func (p *parser) parseDirectivesUntil(shouldStop func(token string) bool) ([]Directive, error) {
-	directives := []Directive{}
-	for {
-		prev := p.tr.lastToken
-		token, ok := p.tr.peek()
-		if !ok {
-			return directives, p.tr.scanner.Err()
-		}
+// Return the next token type without consuming it, or TokenType_EOF if no
+// tokens are left.
+func (p *parser) peekToken() lexer.TokenType {
+	if len(p.tokensLeft) == 0 {
+		return lexer.TokenType_EOF
+	}
+	return p.tokensLeft[0].Type
+}
 
-		if shouldStop(token) {
-			return directives, nil
-		}
-		p.tr.mustConsume(token)
-
-		switch {
-		case strings.HasPrefix(token, "#"):
-			if token == "#" {
-				// `# directive` syntax, read and merge with next token
-				directiveKind, err := p.nextToken()
-				if err != nil {
-					skipped, _ := p.readUntilEOL() // skip remaining part of directive
-					if debug {
-						log.Printf("Failed to parse %v directive: %v, skipping tokens until end of line: %v", token, err, skipped)
-					}
-					break
-				}
-				// parseDirective assumes full directive name including '#' prefix
-				token = "#" + directiveKind
-			}
-			directive, err := p.parseDirective(token)
-			if err != nil {
-				skipped, _ := p.readUntilEOL() // skip remaining part of directive
-				if debug {
-					log.Printf("Failed to parse %v directive: %v, skipping tokens until end of line: %v", token, err, skipped)
-				}
-				break
-			}
-			directives = append(directives, directive)
+// Return the next token and consume it, or TokenEmpty if no tokens are left.
+func (p *parser) nextToken() lexer.Token {
+	if len(p.tokensLeft) == 0 {
+		return lexer.TokenEOF
+	}
 
-		case token == "main":
-			if next, exists := p.tr.next(); exists && next == "(" {
-				if prev == "int" {
-					p.sourceInfo.HasMain = true
+	token := p.tokensLeft[0]
+	p.dropTokens(1)
+	return token
+}
+
+// Check if the next token matches the expected content, returning error
+// otherwise.
+func (p *parser) expectNextToken(expected lexer.TokenType) error {
+	tokenType := p.nextToken().Type
+	switch tokenType {
+	case lexer.TokenType_EOF:
+		return fmt.Errorf("expected %q but reached end of input", expected)
+	case expected:
+		return nil
+	default:
+		return fmt.Errorf("expected %q but found %q", expected, tokenType)
+	}
+}
+
+// parseDirectivesUntil reads tokens and parses directives until shouldStop
+// returns true. It handles main(), #include, and preprocessor blocks, and
+// builds the nested directive structure.
+func (p *parser) parseDirectivesUntil(shouldStop func(token lexer.TokenType) bool) ([]Directive, error) {
+	directives := []Directive{}
+	for !shouldStop(p.peekToken()) {
+		p.parseMainFunction()
```

Comment:
This seems out of place, but perhaps this is just a naming issue? I'd expect `parseMainFunction` to return an error if there's not a `main` function there. Rename to something more tentative like `tryParseMainFunction` or `maybeParseMainFunction`.